Private File Sharding (#44)
wnfs-wg/spec: spec/private-wnfs.md

### 2.1.1 Data Types

> The multimap container MUST be represented as a CBOR-encoded Merkle HAMT. The values MUST be a set of `raw` codec CIDs.

Author: Yeah good point!

> All values in the Merkle HAMT MUST be sorted in binary ascending order by CID and MUST NOT contain duplicates.

Author: Yep. It's not dangerous. In fact `rs-wnfs` would handle that gracefully, I'm pretty sure. It'd just consume the vector and deduplicate it when ingesting it into a `BTreeSet`.

```typescript
type PrivateForest = CBOR<HAMT<Namefilter, Array<CID<ByteArray>>>>

> type HAMT<L, V> = {

Author: Haha. Yeah Rust did. But this is typescript, and even in typescript this is 'normal'. (Just like in Java, etc.)

I *think* I'd prefer it this way for clarity. The `L` is supposed to mean "label", but it should also be evident that it *could* be something else.
I liked the fact that we can describe the private forest as `type PrivateForest = HAMT<Namefilter, Array<CID>>`, for anyone who knows about HAMTs already.
But we could also just remove all the abstraction and generics and make it specific to WNFS.

  structure: "hamt"
  version: "0.1.0"
  root: SparseNode<L, V>
}

type SparseNode<L, V> = [
  ByteArray<2>, // Sparse Index
  Array<Entry<L, V>> // Entries

> ]

Author: This... is supposed to be pseudo-typescript :sweat_smile:
I describe it a little bit in https://github.com/wnfs-wg/spec/blob/main/spec/notation.md.

The spec started out as IPLD schemas. Unfortunately they're not expressive enough for my purposes: All private data would just be `Bytes`. But I want to describe the data that's *inside*. I also *need* to describe the *actual* codec used (dag-cbor) for data that's encrypted.
Right now I manage via the special notation, e.g. `Encrypted<>` and `CBOR<>`, plus comments, but these things just don't exist in IPLD schema land. I've also had some trouble even finding out what the *correct*/most up-to-date IPLD schema language syntax is.

I'd be down for *adding* IPLD schemas to the spec at some point, especially for public WNFS. But I'd like to keep the annotated pseudo-typescript around so we have something that describes how the encryption nests / how things are encoded / what is encrypted with which key.

> ]

Reviewer: > I also need to describe the actual codec used (dag-cbor) for data that's encrypted.

Hmm I definitely see what you're trying to do, but I wonder if we could actually separate these concerns 🤔 I've struggled with the same in the past, but I wonder if instead of handling these at the same layer, we do really clear definitions of the cleartext and ciphertext layers, and show the translation between them via a function call. This means that the spec gets longer, but possibly more clear / easier to follow.

Let's chat about it on our weekly call :)

> ]

Author: In any case - I'm marking this as resolved, as I don't think this has to land in this PR :v:

#### 2.1.1.3 `Entry`

> A leaf node, containing the expanded label as well as a value. As the private forest for WNFS, this MUST contain a namefilter as label and a set of CIDs of ciphertexts of conflicting writes as the value.

Reviewer: Totally a nit, but a suggestion for consistent style:

```suggestion
An `Entry` node MUST contain the following children:

* The expanded label (HAMT hash preimage)
* The one or more CIDs of "values"

The values stored SHOULD be ciphertexts representing conflicting file system writes to that same path and revision.
```

Author: I like it! I'll go for `consists of`, instead of `contains the following children` :)

Author: Ah right, and this is somewhat awkward, since this spec actually defines an abstract `HAMT` data structure and then applies it to `Namefilter`s as keys and `Array<CID>`s as values to and names the applied version the `PrivateForest`.

I'll adjust accordingly. In the future I think we'll probably be able to just 'refer to' the [IPLD HAMT spec](https://github.com/ipld/ipld/tree/master/specs/advanced-data-layouts/hamt).

> A leaf node, containing the expanded label as well as a value. As the private forest for WNFS, this MUST contain a namefilter as label and a set of CIDs of ciphertexts of conflicting writes as the value. This set of CIDs is also called the multi-value.

Author: Yep, I'm aware.
Just trying to introduce terms here. Is that something we should just assume the reader knows? I can remove that section.

> A leaf node, containing the expanded label as well as a value. As the private forest for WNFS, this MUST contain a namefilter as label and a set of CIDs of ciphertexts of conflicting writes as the value. This set of CIDs is also called the multi-value.

Reviewer: Perhaps remove the paragraph and add a link to [multimaps](https://en.wikipedia.org/wiki/Multimap) or similar?

> A leaf node, containing the expanded label as well as a value. As the private forest for WNFS, this MUST contain a namefilter as label and a set of CIDs of ciphertexts of conflicting writes as the value. This set of CIDs is also called the multi-value.

Author: I've changed the wording to "This makes the private forest a mutlimap" with the link to wikipedia.

> A leaf node, containing the expanded label as well as a value. As the private forest for WNFS, this MUST contain a namefilter as label and a set of CIDs of ciphertexts of conflicting writes as the value. This set of CIDs is also called the multi-value.

Author: Actually - I just removed it.

}

type ExternalContent = {
  key: Key
  blockSize: Uint64 // in bytes, at max 262132
  blockCount: Uint64
}
```

### 3.1.1 Node Headers

> Node headers MUST be encrypted with the key derived from the node's skip ratchet: the "content key". Headers MUST NOT grant access to other versions of the associated node. Node headers are in kernel space and MUST NOT be user writable. Refer to the section [Pointers & Keys](#323-pointers--keys) for more detail.

Reviewer: Totally fine as-is, but it also reads well without the words `the section`:

```suggestion
Node headers MUST be encrypted with the key derived from the node's skip ratchet: the "content key". Headers MUST NOT grant access to other versions of the associated node. Node headers are in kernel space and MUST NOT be user writable. Refer to [Pointers & Keys](#323-pointers--keys) for more detail
```

Author: Sounds better! I'm applying the `the section` removal, but I'll keep the period :stuck_out_tongue:

#### 3.2.1.1 Externalized Content

Since external content blocks are separate from the header, they MUST have a unique namefilter derived from a random key (to avoid forcing lookups to go through the header). If the key were derived from the header's key, then the file would be re-encrypted e.g. every time the metadata changed. See [the relevant algorithm](#44-shared-file-content-access) for more detail.

> The block count for externalized content MUST reference the number of blocks the externalized content was split into.

Reviewer: What if it's wrong? i.e. the block claims that there's 100 blocks, but there's only 5

Author: Then at some point reading may fail. Not much we can do about that, right?

We could technically just leave out the count :thinking:, as that's "gossipy", as Brendan likes to say. The end of the file would be reached in case the next block is missing / the first block that's not "full" is hit.

However, what would then happen if a single block is missing in the middle of a file? A client that wants to seek some bytes at the end of the file may read them just fine, but a client that reads from front to back would think it stops in the middle.

Having this gossipy count will let the client reading from front to back know that something went wrong and the file doesn't actually just stop in the middle.

Author: I added some paragraphs about what should happen when it's wrong. Basically file read operations should error out as soon as they notice something is wrong.

> The externalized content's `key` MUST be re-generated randomly every time the file content changes. It MAY stay the same across private file header revisions if only file metadata changes.

Reviewer: ```suggestion
The externalized content's `key` MUST be regenerated randomly whenever the file content changes. If the content stays the same across metadata changes, the content key MAY remain the same across those revisions
```

Reviewer: Will the content sit under both labels, then? If so, can/should we at least change the CID somehow? 🤔 (e.g. prefix with a pepper)

Author: There is only a single label for both revisions in this case. The shard labels only depend on the file node's bare namefilter and `key` for for externalized content, (see [Algorithm 4.4](https://github.com/wnfs-wg/spec/blob/matheus23/file-sharding/spec/private-wnfs.md#44-sharded-file-content-access)).
So if the key stays the same across metadata changes, then all shard labels stay the same, too.

> The namefilters to be used as labels for the ciphertexts in the HAMT are computed as defined in [Algorithm 4.4: Sharded File Content Access](#44-sharded-file-content-access).

Reviewer: ```suggestion
NB: label namefilters MUST be computed per [Algorithm 4.4](#44-sharded-file-content-access)
```

> Each multi-value in the HAMT MUST have exactly one CID. That CID refers to a ciphertext block. Each block MUST be a $2^{18} = 262144$ byte ciphertext, where the first 12 bytes are the initialization vector for encryption, so each block MUST encode exactly 262,132 plaintext bytes, except for the last block with index equal to `blockCount - 1`, which MAY be smaller, but MUST NOT encode an empty plaintext.

Reviewer: ```suggestion
Each HAMT value MUST be referenced by exactly one CID. That CID refers to a ciphertext block. Ciphertexts MUST NOT be black (MUST have at least one byte). A ciphertext chunk MUST NOT exceed $2^{18} = 262144$ bytes of ciphertext, plus a 12-byte initialization vector yielding exactly $262,132$ plaintext bytes.
```

### Chunk Size

I think lifting the forced chunk size may be a good thing, since it would allow us to break chunks on e.g. video frame boundaries.

Determinism would be nice in the general case, but is possibly not _required_ per se. I suggest that we RECOMMEND $2^{18}$ bytes, but let devs lay their data out as needed.

Thoughts?

### Multivalue

I think what you're trying to say is that the each value in the multivalue forms a tree? If so, it's not totally clear from the text as it stands, or in my quick copy edit suggestion above.

![bar](https://user-images.githubusercontent.com/1052016/199845382-d456f02e-17e0-4f2f-a287-ed9211f323df.png)

If so, have you considered making them totally flat, and letting the user derive the entry point via index number? Something like this:

![bar](https://user-images.githubusercontent.com/1052016/199845608-3b6a048b-6327-4153-86c6-d90fcf930157.png)

Where the hash is derived from the index number plus the CID of the first chunk...

$$
\textsf{hash}(index \|| firstCID \|| aesKey)
$$

...or something?

Author: > I think what you're trying to say is that the each value in the multivalue forms a tree?

Ah I guess the wording is pretty bad.
What I was trying to say is that for private file shards, each shard has its own namefilter, and in the HAMT, private shard namefilters must not be multi-value, but have a single value in the multi-value array.

---

> I think lifting the forced chunk size may be a good thing, since it would allow us to break chunks on e.g. video frame boundaries.

Ah, so essentially having a `blockSize` next to `blockCount` in the private file node? That sounds like a great idea. Then the 262,132 bytes just becomes the upper bound on that.

#### 3.1.4.1 Externalized Content

Since external content blocks are separate from the header, they MUST have a unique namefilter derived from a random key (to avoid forcing lookups to go through the header). If the key were derived from the header's key, then the file would be re-encrypted e.g. every time the metadata changed. See [sharded file content access algorithm](#44-shared-file-content-access) for more detail.

> The block size MUST be at least 1 and at maximum $2^{18} - 12 = 262,322$ bytes, as the maximum block size for IPLD is usually $2^{18}$, but 12 initialization vector bytes need to be prepended to each ciphertext. It is RECOMMENDED to use the maximum block size.

Reviewer: Copyediting: run-on sentence

```suggestion
The block MUST NOT be empty, and SHOULD be no larger than $2^{18} - 12 = 262,322$ bytes. This allows both the 12-byte initialization vector and content to fit in the maximum IPFS block size.

Implementations are RECOMMENDED to use the maximum IPLD block size.
```

> The block count MUST reference the number of blocks the externalized content was split into.

Reviewer: What happens if this count is wrong? It this counter here for performance or is it integral to the lookup?

Author: I've tried to answer this question at the end of this section: 

> If any externalized content blocks exceed the specified `blockSize` or are missing in the private forest despite having a lower index than `blockCount` during file read operations, then these operations MUST produce an error.

The idea is to keep this invariant: Two clients should never have conflicting 'ideas' of what the data is. (Admittedly very hand-wavey)
The thinking is this: If we didn't have a count, how would we determine the end of a file?
Probably by reading blocks until we hit the first non-complete or missing block.
This works if we force everyone to start reading files from the beginning, but if someone wants to skip the first 10 blocks and read from then on, they may 'detect' a different end-of-file than someone reading from the beginning, if let's say the 9th block is missing.
I really want to prevent two clients thinking they've detected different file sizes. I'd rather have reads fail when they hit missing blocks. So essentially this makes the block count *the* source of truth.

BFT-ness is really tricky when we're doing things lazily. I guess you can only guarantee BFT-ness if you validate your data-structures deeply, right? I still need to wrap my head around that.

(By the way - you've asked this question the [last time you reviewed this](https://github.com/wnfs-wg/spec/pull/44#discussion_r1013436580) :sweat_smile: )

Reviewer: >  how would we determine the end of a file?

Exponential search!

Author: Why not a count? That's faster. You just know the end of the file.

Exponential search has the same problems I described above.

> The externalized content's `key` MUST be regenerated randomly whenever the file content changes. If the content stays the same across metadata changes, the content key MAY remain the same across those revisions

Reviewer: 👍

NB: Label namefilters MUST be computed as described in the algorithm for [sharded file content access](#44-sharded-file-content-access).

> Entries in the private forest corresponding to externalized content blocks MUST have exactly one CID as their multi-value. This CID MUST refer to a ciphertext with exactly `12 + blockSize` bytes, except for the last block with index `blockCount - 1`. The first 12 bytes refer to the initialization vector and the rest are the ciphertext.

Reviewer: I won't comment on all of them, but stylistically:

```suggestion
Entries in the private forest corresponding to externalized content blocks MUST have exactly one CID as their multi-value. This CID MUST refer to a ciphertext with exactly `12 + blockSize` bytes, except for the last block with index `blockCount - 1`. The first 12 bytes of the block MUST be an initialization vector, and the rest MUST be the ciphertext.
```

> If any externalized content blocks exceed the specified `blockSize` or are missing in the private forest despite having a lower index than `blockCount` during file read operations, then these operations MUST produce an error.

Reviewer: Hmmm, we could consider specifying a minimum number of exceptions for implementations. The exact representation will be different, but we know that this exception is some kind of `BlockSizeExceeded`, exception for example 🤔

Author: I'm not 100% sure whether we should require exceptions to be consistent across implementations, but I'm also not super against it.

I've added this paragraph specifically to address the question of what the source of truth: `blockSize` and `blockCount`are.

### 4.3.2 Seek

> For each path segment, look up the most recent version that can be found (as described in the [private versioning algorithm](#42-Private-Versioning)). This requires access to the directory's revision key.

Reviewer: Nit: an earlier instance referenced it by algorithm number, but this is by lowercase name. I personally like this one more for a few reasons and would change the other one:

* Reads more "naturally"
* You don't have to update all of the section numbers in the links (it's a huge PITA)

##### 4.3.2.1.1 Example

> To illustrate this mode, consider the following diagram. An agent may only have access to some nodes, but not their parent. The agent is able to create new revisions of files and directories, and link that back to the previous version. Some number of revisions may accrue before this can be fully rooted again. Attachment occurs when a reader with enough rights to perform the attachment inspects the file system and discovers that they can attach this file to its parents.

Reviewer: (Wordy)

```suggestion
Consider the following diagram. An agent may only have access to some nodes, but not their parent. The agent is able to create new revisions of files and directories, and link that back to the previous version. Some number of revisions may accrue before this can be fully rooted again. Attachment occurs when a reader with enough rights to perform the attachment inspects the file system and discovers that they can attach this file to its parents.
```

##### 4.3.2.1.1 Example

Consider the following diagram. An agent may only have access to some nodes, but not their parent. The agent is able to create new revisions of files and directories, and link that back to the previous version. Some number of revisions may accrue before this can be fully rooted again. Attachment occurs when a reader with enough rights to perform the attachment inspects the file system and discovers that they can attach this file to its parents.

![](./diagrams/attach.svg)

## 4.4 Sharded File Content Access

`getShards : PrivateFile -> Array<Namefilter>`

> To calculate the array of HAMT labels for [external content](#3211-externalized-content), add `key` and `sha3(key || encode(i))` for each block index `i` of external content to the `bareName` like so: 

Reviewer: 🎉 🧮
